ENH:Add EA types to read CSV

kprestel opened the pull request:
Closes GH23228

- [x ] closes #23228
- [ x] tests added / passed
- [x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x ] whatsnew - May need some  guidance on this part.

kprestel committed:
ENH:Add EA types to read CSV

Closes GH23228

pep8speaks commented:
Hello @kprestel! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/dtypes/common.py`](https://github.com/kprestel/pandas/blob/0151eb0ce6c684a2528b631d87e62d31950392ac/pandas/core/dtypes/common.py) !

 - There are no PEP8 issues in the file [`pandas/tests/io/parser/common.py`](https://github.com/kprestel/pandas/blob/0151eb0ce6c684a2528b631d87e62d31950392ac/pandas/tests/io/parser/common.py) !

kprestel committed:
Address merge comments

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1195,7 +1196,10 @@ cdef class TextReader:
             if result is not None and dtype != 'int64':
-                result = result.astype(dtype)
+                if is_extension_array_dtype(dtype):
+                    result = result.astype(dtype.numpy_dtype)
```
you will need to use 
```
result = dtype.construct_array_type()._from_sequence(result)
```

jreback replied:
may need a try/except around this though

TomAugspurger replied:
We would probably want to raise if the result can't be constructed, right?

kprestel replied:
Yeah I would think so. That would be the defensive thing to do at least. Is that typically what is done in other places?

TomAugspurger replied:
Depends, but in this case I think raising is appropriate.

kprestel replied:
Raising here would actually be a breaking change I think. It broke several tests at least. 

If a user already has an EA class that adheres to the current interface it would all the sudden not work until they defined the new `_from_sequence_of_strings` method. We could try to use the new method and then fallback to the old behavior and warn? 

Not sure how you want to handle this, but I'm pretty sure that it will be a breaking change if we just raised if we couldn't use the new method.

jreback commented on pandas/tests/io/parser/common.py:
```diff
@@ -1603,3 +1604,16 @@ def test_buffer_rd_bytes_bad_unicode(self):
         with pytest.raises(UnicodeError):
             pd.read_csv(t, encoding='UTF-8')
+
+    def test_EA_types(self):
```
we want to try this generically, can you make a new module in pandas/tests/extension/base called io.py and follow the existing patterns.

kprestel replied:
Yeah no problem.

TomAugspurger commented:
Where exactly in the chain of pandas' normal parse, infer, astype, chain does this fall? Is it at the very end?

This seems fragile... Consider a person writing a `DecimalArray` extension type. Would the `DecimalDtype._from_sequence` be given a list of strings? Would pandas have already converted the numeric-like values to floats, thus losing precision? We'll need to better document exactly when this is called.

jreback commented:
see #23595 maybe ``_from_sequence_of_strings`` ? or a more general function

kprestel commented:
@TomAugspurger I'm nowhere near an expert in pandas but from what I can tell the casting happens at the very end. The infer/astype chain seems to start in `parsers.pyx` `self._convert_tokens` where `self == TextReader`. From just reading the code I don't think there is a risk of losing precision as you stated, although it should be tested. 

@jorisvandenbossche Are you suggesting that we add something like `_from_sequence_of_strings` to the EA interface as @jreback suggested?  I can try this and then we can see what it would look like. 

@jreback Can you explain a little bit more about how to create a new test module? I'm not exactly sure what pattern to follow. I'm guessing I need to create a test class and inherit from some base test class but I can't figure out which. Just putting my test in the new `io.py` as you suggested isn't enough because `pytest` doesn't pick it up. Sorry if this is a dumb question, I just haven't worked with a test suite like this before. 

Thanks everyone for their feedback!

TomAugspurger commented:
> From just reading the code I don't think there is a risk of losing precision as you stated, although it should be tested.

As a quick bit of debugging, you could drop a `print(type(scalars[0]))` in `DecimalArray._from_sequence` when you refactor to use that. Though I think @jorisvandenbossche's suggestion of adding a dedicated method to API should be followed.

We should also think about this in the context of https://github.com/pandas-dev/pandas/issues/20612. There will be common patterns for every type of IO we want to add support for (no concrete suggestions right now though).

kprestel commented:
I think whatever this change ends up being it will close #20612. In the original issue (#23228), it is noted that `read_csv` is broken because it will return `object` as the `dtype` if the specified `dtype` is an EA. I believe this is due to the `_get_dtype_type` function in `dtypes/common.py`. 

Basically when you pass a EA to the `dtype` dict in `read_csv` it gets treated as a `type`. Which means `_get_dtype_type` turns it into a `np.object`. 

It seems like that needs to be fixed before anything.

kprestel committed:
WIP:Make python engine support EA types when reading CSVs

The C engine is the real WIP.

jreback commented:
@kprestel pls merge master.

kprestel committed:
Fix CI

kprestel committed:
Formatting

kprestel committed:
Make C engine the same as python parser engine

jreback commented:
can you merge master

kprestel committed:
Merging master

kprestel committed:
Update whatsnew

kprestel commented:
@jreback Done. Sorry about the delay.

kprestel committed:
Fix low_memory C engine parser

kprestel commented:
I think the implementation works and is ready for a full review. I may need a bit of guidance on the testing part though. I've tested it locally and it works but I'm not sure how to test it "more generally" following existing patterns. Any guidance on that aspect would be greatly appreciated. 

Thanks for all the help/time given to me so far!

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -26,6 +26,8 @@ New features
 - :meth:`DataFrame.corr` and :meth:`Series.corr` now accept a callable for generic calculation methods of correlation, e.g. histogram intersection (:issue:`22684`)
 - :func:`DataFrame.to_string` now accepts ``decimal`` as an argument, allowing the user to specify which decimal separator should be used in the output. (:issue:`23614`)
 - :func:`DataFrame.read_feather` now accepts ``columns`` as an argument, allowing the user to specify which columns should be read. (:issue:`24025`)
+- :func:`pandas.read_csv` now supports ``EA`` types as an argument to ``dtype``,
```
instead of EA types, say pandas extension dtypes.

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -885,12 +888,14 @@ cdef class TextReader:
         cdef:
             int status
 
+        print(f'low mem: {self.low_memory}')
```
remove print statements

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1229,7 +1238,20 @@ cdef class TextReader:
             if result is not None and dtype != 'int64':
-                result = result.astype(dtype)
+                if is_extension_array_dtype(dtype):
+                    try:
```
why do you have a try/except here? is theere a test that fails? this is a requirements for EA dtypes.

kprestel replied:
If it was a string, but it shouldn't ever be. Taking that out.

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1229,7 +1238,20 @@ cdef class TextReader:
 
             if result is not None and dtype != 'int64':
-                result = result.astype(dtype)
+                if is_extension_array_dtype(dtype):
```
this ``if is_extension_array_dtype`` should be right after the ``is_categorical_dtye``, IOW on line 1227.

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1229,7 +1238,20 @@ cdef class TextReader:
+                        # use _from_sequence_of_strings if the class defines it
+                        result = array_type._from_sequence_of_strings(result,
+                                                                      dtype=dtype) # noqa
+                    except AbstractMethodError:
```
don't use AbstractMethodError, rather catch NotImplementedError

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1238,9 +1260,21 @@ cdef class TextReader:
 
             if result is not None and dtype != 'float64':
-                result = result.astype(dtype)
+                if is_extension_array_dtype(dtype):
```
why are you repeating?

kprestel replied:
bad code... changing where the `is_extension_array_dtype` makes this unnecessary anyway.

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -2173,7 +2207,11 @@ def _concatenate_chunks(list chunks):
                                               sort_categories=sort_categories)
         else:
-            result[name] = np.concatenate(arrs)
+            if is_extension_array_dtype(dtype):
```
what is this for?

kprestel replied:
without this the `np.concatenate` call will return the `arrs` back to an `np.object_` array again.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -123,6 +123,27 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
+    @classmethod
+    def _from_sequence_of_strings(cls, strings, dtype=None, copy=False):
+        """Construct a new ExtensionArray from a sequence of scalars.
+
```
add a versionadded here

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -154,7 +154,10 @@ def coerce_to_array(values, dtype, mask=None, copy=False):
         if not issubclass(type(dtype), _IntegerDtype):
             try:
-                dtype = _dtypes[str(np.dtype(dtype))]
+                try:
```
why are you changing this?

kprestel replied:
if `dtype` is a class (not instantiated) then it's `type(dtype) == type` making this statement true, but then a `KeyError` gets raised because it can't find the `dtype` in the `_dtypes` dict. 

I know this isn't the best solution but I can't think of a better one. If you have a better solution then I would be happy to implement it.

jorisvandenbossche replied:
I think we can require that `dtype` is an instantiated object and not a class?, or where does this come up?

kprestel replied:
I don't think requiring `dtype` to be instantiated is the correct behavior. The following test, which is based on the original issue, will fail with the error message: 
`ValueError: invalid dtype specified <class 'pandas.core.arrays.integer.Int64Dtype'>`

```python
@pytest.mark.parametrize('engine', ['c', 'python'])
def test_EA_types(engine):
    df = pd.DataFrame({'Int': pd.Series([1, 2, 3], dtype='Int64'),
                       'A': [1, 2, 1]})
    data = df.to_csv(index=False)
    result = pd.read_csv(StringIO(data), dtype={'Int': Int64Dtype},
                         engine=engine)
```

I guess we could insatiate the `dtype` somewhere further up in the call stack, but to me this seems like the correct place to do it. This is where we turn `str` objects into the correct `dtype`, so it makes sense that we handle class references here as well.

jreback replied:
this should not be changed

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -123,6 +123,27 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
+        Parameters
+        ----------
+        strings : Sequence
+            Each element will be an instance of the scalar type for this
```
IIRC this came up in the discussion of Interval parsing from strings. did we already create some machinery fro this? @jschendel @TomAugspurger @jorisvandenbossche

jorisvandenbossche replied:
I am not fully sure what you are hinting on. But, the docstring is certainly wrong here (I suppose copy pasted from `_from_sequence`), as here the input should be a sequence of strings, not of scalars of the dtype.

kprestel replied:
Fixed the docstring.

I wasn't able to find anything that added the machinery, but if I'm wrong let me know and I will use it instead of this new method.

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -261,6 +264,10 @@ def __init__(self, values, mask, copy=False):
 
+    @classmethod
+    def _from_sequence_of_strings(cls, strings, dtype=None, copy=False):
+        return cls._from_sequence([int(x) for x in strings], dtype, copy)
```
this is not going to work, what if you have a NaN? I think instead you need to call
``pd.to_numeric(errors='raise')`` on this, then pass to cls._from_sequence

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -615,8 +616,27 @@ def astype_nansafe(arr, dtype, copy=True, skipna=False):
-        return dtype.construct_array_type()._from_sequence(
-            arr, dtype=dtype, copy=copy)
+        if is_object_dtype(arr):
+            try:
```
woa, what is this all about?

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -1795,7 +1795,10 @@ def _get_dtype(arr_or_dtype):
         return arr_or_dtype
     elif isinstance(arr_or_dtype, type):
-        return np.dtype(arr_or_dtype)
+        try:
```
what are these changes for?

kprestel replied:
Thought I needed them at some point. Tests seem to pass without them. Reverted.

jreback commented on pandas/tests/extension/base/io.py:
```diff
@@ -0,0 +1,44 @@
+        df = pd.DataFrame({'Int': pd.Series([1, 2, 3], dtype='Int8'),
+                           'A': [1, 2, 1]})
+        data = df.to_csv(index=False)
+        result = pd.read_csv(StringIO(data), dtype={'Int': 'Int8'},
```
this needs to be generic, pass in in ``str(data.dtype)`` here

jreback reviewed (changes requested):
you are having WAY too many catching of dtypes. Rip all of that out and just leave the code in parser.pyx

kprestel commented:
@jreback You're right and not 20 minutes after I pushed it did I think about that... I was too focused on the use case you described in the original issue, however if I just rip out everything that isn't in `parsers.pyx` I don't think it would work at all. If it did it wouldn't work the python parser for sure. 

I will reply to your comments and answer your questions.

Thanks for the review. I really appreciate it.

jreback commented:
@kprestel for > https://github.com/pandas-dev/pandas/pull/23255#issuecomment-445563694

right, the python parser will need some small fixes as well. but let's start minimally.

kprestel committed:
Address merge comments

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -123,6 +123,30 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
+        ExtensionArray
+
+        """
+        raise AbstractMethodError(cls)
```
I would make this a NotImplementedError, as I don't think it should be required to implement this?

jorisvandenbossche commented on pandas/_libs/parsers.pyx:
```diff
@@ -1215,6 +1215,18 @@ cdef class TextReader:
+                              ._from_sequence_of_strings(result, dtype=dtype)
+            except NotImplementedError:
+                result = dtype.construct_array_type() \
+                              ._from_sequence(result, dtype=dtype)
```
Rather than falling back, we could also raise an informative error message? (or simply let the NotImplementedError through to the user, assuming this already has a nice error message)

jorisvandenbossche commented on pandas/core/arrays/integer.py:
```diff
@@ -261,6 +265,11 @@ def __init__(self, values, mask, copy=False):
 
+    @classmethod
+    def _from_sequence_of_strings(cls, strings, dtype=None, copy=False):
+        scalars = to_numeric(strings, errors='raise')
```
we need to ensure here that the result are integers I think? (or raise an error if that is not the case)

kprestel replied:
I believe that between the `errors='raise'` and the subsequent call to `cls._from_sequence` would raise the appropriate error.

jorisvandenbossche commented on pandas/core/dtypes/cast.py:
```diff
@@ -615,8 +616,18 @@ def astype_nansafe(arr, dtype, copy=True, skipna=False):
+            array_type = dtype.construct_array_type()
+        try:
+            # use _from_sequence_of_strings if the class defines it
+            return array_type._from_sequence_of_strings(arr,
```
why is `_from_sequence_of_strings` being used here? As this is for general casting and not necessarily string parsing?

kprestel replied:
It is used by the `python` engine, `PythonParser._cast_types`. When it gets to this point it is casting the strings (from the CSV) into `ndarrays`.

jreback replied:
why is this changed at all?

jreback replied:
@kprestel don't resolve something unless you respond. you haven't answered the question at all.

kprestel committed:
Require EAs to implement _from_sequence_of_strings to be used in parsers

kprestel committed:
pep8

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1215,6 +1215,22 @@ cdef class TextReader:
+                                                    na_hashset)
+            try:
+                # use _from_sequence_of_strings if the class defines it
+                result = dtype.construct_array_type() \
```
use
```
array_type = dtype.construct_array_type()
result = array_type._from...
```
to avoid this line-continuation (harder to read)

jreback replied:
the array_type can be outside the try/except

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1215,6 +1215,22 @@ cdef class TextReader:
+                    "Extension Array: {ea} must implement "
+                    "_from_sequence_of_strings in order "
+                    "to be used in parser methods".format(
+                            ea=dtype.construct_array_type()))
```
you can use the array_type from above

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -2173,7 +2188,11 @@ def _concatenate_chunks(list chunks):
         else:
-            result[name] = np.concatenate(arrs)
+            if is_extension_array_dtype(dtype):
+                result[name] = dtype \
```
same as above, avoid line continuation

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -123,6 +123,30 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
+        ExtensionArray
+
+        """
+        raise NotImplementedError(cls)
```
use AbstractMethodErrror

jreback reviewed (changes requested):
looking closer, but you should not need to modify antyhing in pandas/core/dtypes/cast.py if you think you do, then show me exactly why with a self-contained example.

kprestel commented:
@jreback You do need the changes in `pandas/core/dtypes/cast.py`. 

Here is an example:

```python
import pandas as pd
from pandas.compat import StringIO
from pandas.core.arrays.integer import Int64Dtype

df = pd.DataFrame({'Int': pd.Series([1, 2, 3], dtype='Int64'),
                   'A': [1, 2, 1]})
data = df.to_csv(index=False)
# this will raise a TypeError: object cannot be converted to IntegerDtype
result = pd.read_csv(StringIO(data), dtype={'Int': Int64Dtype},
                     engine='python')
# this will raise a AttributeError: 'str' object has no attribute 'construct_array_type'
result = pd.read_csv(StringIO(data), dtype={'Int': 'Int64'},
                     engine='python')
```

This example assumes you are on my branch and have reverted my changes to `pandas/core/dtypes/cast.py`


Using this same example code, if you were to revert the changes on line 158 in `pandas/core/arrays/integer.py` the first example using `Int64Dtype` will fail with this exception: 
`ValueError: invalid dtype specified <class 'pandas.core.arrays.integer.Int64Dtype'>`

This is the example from the original issue. I replied to that comment with a more detailed explanation as to why I think it should stay there, if you still think I should pull it out I will and update the necessary test cases.

kprestel committed:
Address merge comments

Reverted unintended changes to whatsnew

Raise AbstractMethodError instead of NotImplemented

Clean up changes in parsers.pyx

jreback commented:
@kprestel not what I mean. you certainly need a change in ``parsers.py`` to handle dtype inference for EA for the python parser. But I don't think you need *any* changes to the actual dtype routines themselves. I would need an example that is on master.

kprestel commented:
@jreback ah, ok. I see what you mean. Sorry about the misunderstanding. But just to make sure I'm clear, you're saying what I'm doing (inferring the dtype for EAs) is correct, I'm just not doing it in the correct place? It should be done in `parsers .py`.

jreback commented:
> @jreback ah, ok. I see what you mean. Sorry about the misunderstanding. But just to make sure I'm clear, you're saying what I'm doing (inferring the dtype for EAs) is correct, I'm just not doing it in the correct place? It should be done in parsers .py.

for the engine='c' case its fine, you just need a path for teh python parser

kprestel committed:
Do EA type inference in the parser method rather than the cast method

kprestel commented:
Ok, I made the python parser work more like the `C` parser. Should have thought to do that to begin with... Sorry about that.

kprestel committed:
Fix imports using isort

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -50,13 +50,14 @@ from pandas.core.dtypes.common import (
 
-from pandas.errors import (ParserError, DtypeWarning,
-                           EmptyDataError, ParserWarning)
+from pandas.errors import ( ParserError, DtypeWarning,
```
extra spaces here

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -154,7 +155,10 @@ def coerce_to_array(values, dtype, mask=None, copy=False):
             try:
-                dtype = _dtypes[str(np.dtype(dtype))]
+                try:
+                    dtype = _dtypes[str(np.dtype(dtype.name.lower()))]
```
why is this changed?

jreback commented on pandas/core/indexes/numeric.py:
```diff
@@ -328,7 +328,8 @@ def astype(self, dtype, copy=True):
                    'values are required for conversion').format(dtype=dtype)
             raise TypeError(msg)
-        elif is_integer_dtype(dtype) and self.hasnans:
+        elif is_integer_dtype(dtype) and self.hasnans and \
```
why is this needed?

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -558,7 +558,8 @@ def sanitize_array(data, index, dtype=None, copy=False,
 
             # possibility of nan -> garbage
-            if is_float_dtype(data.dtype) and is_integer_dtype(dtype):
+            if is_float_dtype(data.dtype) and is_integer_dtype(dtype) \
```
why is this needed?

jreback commented on pandas/io/parsers.py:
```diff
@@ -1660,15 +1660,17 @@ def _convert_to_ndarrays(self, dct, na_values, na_fvalues, verbose=False,
             else:
                 # skip inference if specified dtype is object
-                try_num_bool = not (cast_type and is_string_dtype(cast_type))
+                try_num_bool = not (cast_type and (is_string_dtype(cast_type)
```
don't use a noqa, rather make another variable this condition is hard to read / understand

jreback commented on pandas/io/parsers.py:
```diff
@@ -1660,15 +1660,17 @@ def _convert_to_ndarrays(self, dct, na_values, na_fvalues, verbose=False,
 
                 # type specified in dtype param
-                if cast_type and not is_dtype_equal(cvals, cast_type):
+                if cast_type and (not is_dtype_equal(cvals, cast_type)
```
can you add to the comment

jreback commented on pandas/io/parsers.py:
```diff
@@ -1765,6 +1767,23 @@ def _cast_types(self, values, cast_type, column):
                 cats, cats.get_indexer(values), cast_type,
                 true_values=self.true_values)
 
+        elif is_extension_array_dtype(cast_type):
```
can you comment here

jreback commented on pandas/tests/extension/base/io.py:
```diff
@@ -0,0 +1,39 @@
+
+def make_data():
+    return (list(range(1, 9)) + [np.nan] + list(range(10, 98))
+            + [np.nan] + [99, 100])
```
you don't need *any* fixtures here

jreback commented on pandas/tests/extension/base/io.py:
```diff
@@ -0,0 +1,39 @@
+        df = pd.DataFrame({'Int': pd.Series(data, dtype=str(data.dtype)),
+                           'A': data})
+        data = df.to_csv(index=False)
+        result = pd.read_csv(StringIO(data), dtype={'Int': str(data.dtype)},
```
make this a bit more generic (don't call it Int)

jreback commented on pandas/tests/extension/test_common.py:
```diff
@@ -1,14 +1,35 @@
 
 
+def make_data():
+    return (list(range(1, 9)) + [np.nan] + list(range(10, 98))
```
why are you adding fixtures here?

kprestel replied:
this file shouldn't have been committed, I was testing things locally and forgot that I shouldn't have added tests here.

jreback reviewed (changes requested):
for the tests, you just need to add your new testing extension to *each* of the tests for the extensions, e.g.

```
class TestParsing(base.BaseParsingTests):
    pass
```

in pandas/tests/extensions/*

kprestel committed:
Correct tests to match existing patterns, address merge comments

Ensure that dtype is always an initiated object

kprestel commented:
@jreback I updated the tests based on what you said, let me know if it is still wrong. 

I also resolved your comments, most of them had to do with handling uninstantiated EAs. My change to `pandas_dtypes` resolved all of this, so I pulled all of the extra checks out. 

Thanks again for the feedback, I really appreciate it.

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -31,7 +31,10 @@ New features
+- :func:`pandas.read_csv` now supports pandas extension types as an argument to ``dtype``, allowing the user to use pandas extension types when reading CSVs. (:issue:`23228`)
 - :meth:`DataFrame.shift` :meth:`Series.shift`, :meth:`ExtensionArray.shift`, :meth:`SparseArray.shift`, :meth:`Period.shift`, :meth:`GroupBy.shift`, :meth:`Categorical.shift`, :meth:`NDFrame.shift` and :meth:`Block.shift` now accept `fill_value` as an argument, allowing the user to specify a value which will be used instead of NA/NaT in the empty periods. (:issue:`15486`)
+- :func:`pandas.read_csv` now supports ``EA`` types as an argument to ``dtype``,
+  allowing the user to use ``EA`` types when reading CSVs. (:issue:`23228`)
```
looks like the note is duplicated? the first one looks good, so remove this one

jreback commented on pandas/_libs/parsers.pyx:
```diff
@@ -1123,7 +1122,9 @@ cdef class TextReader:
 
-            if upcast_na and na_count > 0:
+            try_upcast = upcast_na and na_count > 0
+            # don't try to upcast EAs
```
can you move this comment above the ``try_upcast = ``

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -1986,6 +1986,9 @@ def pandas_dtype(dtype):
     result = _pandas_registry.find(dtype) or registry.find(dtype)
     if result is not None:
+        # ensure result is an instantiated type
+        if isinstance(result, type):
```
can you explain where exactly you think this needed? this violates the return values of this function and should not be needed here. This function returns a type not an instance

jreback replied:
ok great!

kprestel replied:
One place that this did come up is in `is_bool_dtype` though. I think there a bug there. I can open an issue about it if you want, but basically what happens is if an a EA class is passed in it will return a property method. This can create unexpected side effects. for example:

```python
if is_bool_dtype(Int64dtype):
    print('is a bool dtype')
else:
    print('not a bool dtype')
```

This will print `is a bool dtype` because `is_bool_dtype` won't return `True' or `False`, it will actually return a `property_method` reference. 

That is (one of) the reasons I thought that I needed this.

jreback commented on pandas/io/parsers.py:
```diff
@@ -1765,6 +1769,23 @@ def _cast_types(self, values, cast_type, column):
+            try:
+                array_type = cast_type.construct_array_type()
+            except AttributeError:
+                cast_type = pandas_dtype(cast_type)
```
why are you catching AttributeError here? this should *always* exist, it is a required feature of EAs (if it doesn't then we need to fix that).

What is this failing on ?

kprestel replied:
Don't need this. It was before it was in this `if` block, just didn't remove it.

jreback commented on pandas/io/parsers.py:
```diff
@@ -2174,7 +2195,14 @@ def __init__(self, f, **kwds):
         self.verbose = kwds['verbose']
         self.converters = kwds['converters']
-        self.dtype = kwds['dtype']
+
```
this shouldn't be done here, rather you can do this if needed inside the parser itself

jreback commented on pandas/tests/extension/base/io.py:
```diff
@@ -0,0 +1,24 @@
+            'with_dtype': str(data.dtype)
+        }, engine=engine)
+        assert result is not None
+        tm.assert_frame_equal(df, result)
```
you don't need the ``assert result is not None``

```
expected = df
tm.assert_frame_equal(result, expected)
```

jreback commented on pandas/tests/extension/test_interval.py:
```diff
@@ -152,3 +152,11 @@ class TestPrinting(BaseInterval, base.BasePrintingTests):
         pass
+
+
+class TestParsing(BaseInterval, base.BaseParsingTests):
```
great!

jreback reviewed (changes requested):
@kprestel looking really good. Just a couple of comments. You are working on a pretty intricate and wide ranging code here, but close!

kprestel committed:
Addressing more merge comments

Reverted change where pandas_dtype returns instantiated object

kprestel commented:
The CI build should be fixed now. You're right changing `pandas_dtype` was a bad idea...

jreback commented:
lgtm. @TomAugspurger @gfyoung any comments.

kprestel commented:
@jreback Thanks again for all the help along the way! I really appreciated it.

jreback commented:
@kprestel no, thank you! this is a really useful feature!

actually now that I think about it. Can you add an example in io.rst (and maybe in the read_csv doc-string) showing say reading in as Int64 (I think we already have an example, so you can just add another column that is Int64)

kprestel committed:
Update docs to show using Int64 dtype

kprestel commented:
@jreback Done. Let me know if that isn't what you were looking for.

jreback commented:
perfect

gfyoung commented on pandas/tests/extension/base/io.py:
```diff
@@ -0,0 +1,24 @@
+            'with_dtype': str(data.dtype)
+        }, engine=engine)
+        expected = df
+        tm.assert_frame_equal(result, expected)
```
Hmm...I really would want to somehow take this test and plant it in the `parsers` tests.

That being said, it's nicely integrated with the existing extension tests that I'm not sure it's worth it...

Perhaps we should add a small test with EA types in the `parsers/common.py` ?

kprestel replied:
Honestly I don't think it would add much value. This feature needs to be tested generically. When I was developing this feature had a small test that I was using but it was too specific to really add any value and didn't cover any use cases that we not already covered by the existing test suite. 

I can add a small test there if you still want, but IMHO it won't add that much value.

jreback replied:
yeah this is ok here; @gfyoung  wouldn't object to *also* having a test in parsers/common.py that maybe reads in a sample with a bunch of EA types (thinking Decimal and Int64), integrated with the existing tests. but can be a followup.

gfyoung replied:
> wouldn't object to also having a test in parsers/common.py that maybe reads in a sample with a bunch of EA types (thinking Decimal and Int64), integrated with the existing tests. but can be a followup.

@jreback : Yep, that's what I had in mind.

jreback replied:
thanks @gfyoung

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -128,6 +128,30 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
         """
         raise AbstractMethodError(cls)
 
+    @classmethod
```
can you add ``_from_sequence_of_strings`` in the Notes section at the top, next to the other constructors

jreback replied:
actually this is no required, so put it separately

kprestel committed:
Update docs per comments in PR

jreback commented:
@kprestel some lint issues:https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=6210

jreback commented:
ping on green.

kprestel committed:
Fix linters

kprestel commented:
Seems like this needs to get updated?

2018-12-30T20:24:59.3212208Z Check for invalid EA testing
2018-12-30T20:24:59.3294654Z ##[error]pandas/tests/extension/base/io.py(24,): error :  Found unwanted pattern:         tm.assert_frame_equal(result, expected)

Otherwise it should be fine now.

jreback commented on pandas/tests/extension/base/io.py:
```diff
@@ -0,0 +1,24 @@
+            'with_dtype': str(data.dtype)
+        }, engine=engine)
+        expected = df
+        tm.assert_frame_equal(result, expected)
```
right here use ``self.assert_frame_equal``. This is to allow more specific testing of EA types.

kprestel replied:
ah. thanks.

kprestel committed:
Fix linters again

kprestel committed:
isort

kprestel commented:
@jreback should be good after this build. 

Sorry for the delay.

jreback commented:
thanks @kprestel we have pretty strict ci / linters now-a-days; a good thing.

jreback commented:
thanks @kprestel really nice!

jreback commented:
keep em coming!

kprestel commented:
I plan on it. Thanks again for the help!
